### DISCUSSION TITLE
Dotard and Simbleton chapter 13 Find()

### DISCUSSION THREAD

**Discussion by Chadwell-Clark** [nss-day-cohort-47/C47-Discussions#24] [Help Me] [answered]
I have not been able to get this to work yet and am probably not implementing it correctly.
Just wondering  if others have figured it out and/or if anyone would want to discuss it or walk through it. I would like to spend more time on these array methods as they are powerful tools and want to get their syntax down. I have scoured the gurgle, tubes and have been overflowed with stack but most examples of find are simple "find the first occurrence of an array string" or not relevant to how it is being used. Interested in figuring this out. Hard for me to not to complete everything I can. Would welcome fellow travelers. <-/^\\->

[Chapter 13 Dotard Simbleton](https://github.com/nss-day-cohort-47/client-side-mastery/blob/master/book-1-martins-aquarium/chapters/DS_FIND_COMPANY.md)

**Comment by ardisrtilson** [comment]
Just a hint: .find() indeed simply finds and returns the first occurrence of a string or value in an array which matches a parameter entered by the programmer. By itself, this seems like a minor function, but in the real world these array methods are usually "nested" within each other or "chained" in a sequence to perform more complex tasks. The most common nest / chain pattern you'll see is:

return ourArray.map((ourArrayValue) => {
         return ourObjects.find(OurObjectValue => OurObjectValue.value === ourArrayValue)
});

If this looks like gobbledegook don't worry, we wont be here for a few weeks. It's just to show how array methods are usually used in more complex code.

**Reply by Chadwell-Clark** [reply]
It does not look like gobbledegook. Perhaps I am wrong but it seems map is being to used return an array of objects that will either be undefined or have passed the conditional argument of find. If I am completely in left field I would love to get it explained.

**Comment by BryanNilsen** [comment, marked as answer]
I think it's important to think about everything in context first and use regular words...
**.find()** = I have this collection of data and I want to find this **one thing** that matches a certain condition or set of conditions.
**.filter()** =  I have this collection of data and I want to find **all of the things** that matches a certain condition or set of conditions.

let's look at **.find()** - returns the first object that meets a specific condition
```
const colors = [ "orange", "blue", "yellow", "green", "red", "black", "purple", "white"]

// Find the first item in the array that starts with the letter 'b'

const firstColorThatStartsWithB = colors.find(color => color.startsWith('b'))
```
1. start with the array:
`colors`

2. add the array method:
`colors.find()`

3. give a variable name to each item in the array and add the arrow syntax:
`colors.find(color =>)`

4. add a statement (or condition) that will return a boolean (true/false):
colors.find(color => color.startsWith('b'))

5. store the result in a new variable:
`const firstColorThatStartsWithB = colors.find(color => color.startsWith('b'))`


let's look at **.filter()** - syntax is the same, although the result is a new array with the items that match the condition
```
// Find all items in the array that start with the letter 'b'

const allColorsThatStartsWithB = colors.filter(color => color.startsWith('b'))
```

This is a real simplistic overview for right now as far as breaking down the syntax that you'll most likely be seeing at first. We can dive in deeper when you're ready!

**Reply by Chadwell-Clark** [reply]
I do understand find() in this context. My issue is that I am not able to use it to capture the input of the search in Dotard Simbleton. I tried using the devtools to see what is going on upon pressing return but it just sees it as a keyboard event of enter. 

I have this code, most of which is from the chapter. I changed a few things because **keypress** and **charCode** are **deprecated** I feel the issue a matter of implementation of the input value. These are the types of things that bug me to no end and I just can't let go of. I have spent too much time on this already. Way too much. I *need* to **find** the Way

I am calling finder() in main after importing it. I am not getting any errors and the DOM is not broken. 
```
export const finder = () => {
const companySearchResultArticle = document.querySelector(".foundCompanies")

document
    .querySelector("#companySearch")
    .addEventListener("keydown", event => {
        if (event.key === 13) {
            /*
                When user presses enter, find the matching business.
                You can use the `.includes()` method strings to
                see if a smaller string is part of a larger string.

                Example:
                    business.companyName.includes(keyPressEvent.target.value)
            */

            const foundBusiness = getBusinesses().find(business => business.companyName.startsWith(event.target.value)) //?? I feel like I am not targeting the right thing

            companySearchResultArticle.innerHTML = `
                <h2>
                ${foundBusiness.companyName}
                </h2>
                <section>
                ${foundBusiness.addressFullStreet}

                </section>
                <section>
                ${foundBusiness.addressCity},
                ${foundBusiness.addressStateCode}
                ${foundBusiness.addressZipCode}
                </section>
            `;
        }
    });
}
```

**Reply by Chadwell-Clark** [reply]
I am guessing that I don't need to export it as it is listening and has the array it needs to do the search and the HTML is ready to be inserted in the DOM, well as soon as the find() finds properly.

**Reply by Chadwell-Clark** [reply]
This morning I changed (event.key === 13) to (event.key === "Enter").
I was in the shower when it came to me from the  swirling steam, drain or shower light. 
This did make the first match show up. It did not format properly but it did show up.

**Reply by Chadwell-Clark** [reply]
Nothing big but I also noticed in Chapter 13 it shows:

## Videos With More Examples

You can watch some videos where other developers show you other examples of how to use the `.map()` method on an array.   ***&larr;//Not sure if map() is meant to be find() or not***

* [ES6 Tutorial - how to use javascript array FIND method](https://www.youtube.com/watch?v=PMRkx51LGSQ)
* [8 Must Know JavaScript Array Methods](https://youtu.be/R8rmfD9Y5-c?t=174)

**Reply by BryanNilsen** [reply]
`const foundBusiness = getBusinesses().find(business => business.companyName.startsWith(event.target.value))`
.find() will return only the first match.
if you want more than one result, use a .filter()
👍

**Reply by Chadwell-Clark** [reply]
# Green Answered Checkmark!
I probably should have put this in help so i could give you an answered check mark. 

Found it!!

**Reply by EthanMathis** [reply]
Chad you're a saint and a scholar. I was throwing myself at this problem to no avail lol. If it weren't for you asking this ^ i never would've figured this out on my own. Thanks :)

**Reply by Chadwell-Clark** [reply]
I don't know about all that scholarly saint stuff... but I am glad others can learn from my mistakes. I will make more so we can all learn!
